File: javascript/smartlist/smartlist-api-mysql-project/src/modules/candidate/candidate.repository.ts
import { EntityRepository, Repository } from 'typeorm';

import { Injectable } from '@nestjs/common';
import { Candidate } from 'src/entities/candidate.entity';
import { Organization } from 'src/entities/organization.entity';

@Injectable()
@EntityRepository(Candidate)
export class CandidateRepository extends Repository<Candidate> {
  constructor() {
    super();
  }

  async findById(id: number): Promise<Candidate> {
    const query = {
      where: { id: id },
      relations: ['photo'],
    };

    return this.findOne(query);
  }

  async findByContactNumber(contactNumber: string): Promise<Candidate> {
    return this.findOne({
      where: { contactNumber: contactNumber },
      relations: ['photo'],
    });
  }

  async findByEmail(email: string): Promise<Candidate> {
    return this.findOne({
      where: { email: email },
    });
  }

  async findByNameForOrganization(
    organization: Organization,
    name: string,
  ): Promise<Candidate[]> {
    return this.createQueryBuilder('candidate')
      .select('candidate')
      .leftJoinAndSelect('candidate.photo', 'photo')
      .innerJoin('candidate.candidateAssessments', 'candidateAssessments')
      .innerJoin('candidateAssessments.assessment', 'assessment')
      .innerJoin('assessment.organization', 'organization')
      .where('organization.id = :id', { id: organization.id })
      .andWhere('candidate.name like :name', { name: '%' + name + '%' })
      .distinctOn(['candidate.id'])
      .getMany();
  }
}

File: java/smartlist/src/main/java/com/smartlist/module/candidate/CandidateRepository.java
package com.smartlist.module.candidate;

import com.smartlist.model.Candidate;
import com.smartlist.model.Organization;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface CandidateRepository extends CrudRepository<Candidate, Integer> {

    Optional<Candidate> findByContactNumber(String contactNumber);

    @Query("SELECT c FROM Candidate c JOIN c.candidateAssessments ca WHERE ca.assessment.organization = :organization AND c.name LIKE %:name%")
    List<Candidate> findByNameForOrganization(@Param("name") String name, Organization organization);

    Optional<Candidate> findByEmail(String email);
}
